test/config-luatest/config_include_test.lua

```diff
@@ -0,0 +1,621 @@
+                                },
+                            },
+                        },
+                    },
+                },
+            },
+        },
+    }
+
+    treegen.write_file(
+        g.dir, 'conf.d/included_config_1.yaml', yaml.encode(included_config_1))
+    treegen.write_file(
+        g.dir, 'conf.d/included_config_2.yaml', yaml.encode(included_config_2))
+
+    local opts = {
+        config_file = config_file,
+        alias = 'instance-001',
+        chdir = g.dir,
+    }
+    g.server = server:new(opts)
+    g.server:start()
+
+    g.server:exec(function()
+        t.assert_equals(
+            require('config'):get('labels'),
+            {
+                value_1 = 'from config.yaml',
+                value_2 = 'from included_config_2.yaml',
+                value_3 = 'from included_config_1.yaml',
+            }
+        )
+    end)
+end
+
+g.test_include_recursion = function()
+    local config = table.deepcopy(common_config)
+    config.log = {to = 'file'}
+
+    local included_config = {
+        include = {'config.yaml'},
+        groups = {
+            ['group-001'] = {
+                replicasets = {
+                    ['replicaset-001'] = {
+                        instances = {
+                            ['instance-001'] = {
+                                labels = {
+                                    value_2 = 'from included_config.yaml',
+                                    value_3 = 'from included_config.yaml',
+                                },
+                            },
+                        },
+                    },
+                },
+            },
+        },
+    }
+
+    local config_file = treegen.write_file(
+        g.dir, 'config.yaml', yaml.encode(config))
+    treegen.write_file(
+        g.dir, 'included_config.yaml', yaml.encode(included_config))
+
+    local opts = {
+        config_file = config_file,
+        alias = 'instance-001',
+        chdir = g.dir,
+    }
+    g.server = server:new(opts)
+    g.server:start()
+
+    g.server:exec(function()
+        t.assert_equals(
+            require('config'):get('labels'),
+            {
+                value_1 = 'from config.yaml',
+                value_2 = 'from config.yaml',
+                value_3 = nil,
+            }
+        )
+    end)
+
+    config.include = {'included_config.yaml'}
+    treegen.write_file(g.dir, 'config.yaml', yaml.encode(config))
+
+    g.server:exec(function()
+        require('config'):reload()
+        t.assert_equals(
+            require('config'):get('labels'),
+            {
+                value_1 = 'from config.yaml',
+                value_2 = 'from included_config.yaml',
+                value_3 = 'from included_config.yaml',
+            }
+        )
+    end)
+
+    g.server:grep_log(
+        ('skipping already processed config file: ' ..
+        '%s/config.yaml"'):format(g.dir),
+        1024, {filename = g.dir .. '/var/log/instance-001/tarantool.log'}
+    )
```

Comment:
I'm in some fear about `:grep_log()` without any retrying. A frequent source of flaky tests.

At least until https://github.com/tarantool/luatest/pull/422 goes to tarantool (that should go with https://github.com/tarantool/test-run/pull/455, but it has some problems AFAIR and stalled so).

Reply:
Ok, good take. I've added `retrying` helper.

Reply:
It seems correct, but I overlooked another mistake past time: `:grep_log()` returns `nil` if the pattern is not found. It doesn't raise an error on its own. I've looked over existing tests and it seems that the usual pattern is `t.assert(server:grep_log(<...>))`.

Reply:
yeap, missed it, thanks. Fixed.